src/shared/components/table/state/tableReducer.ts

```diff
@@ -0,0 +1,155 @@
+ */
+import type { ReactNode } from 'react';
+
+enum SortingDirection {
+  ASC = 'ascending',
+  DESC = 'descending',
+  DEFAULT = 'default' // Do we need this?
+}
+
+export type SortedColumn = {
+  columnId: string;
+  sortedDirection: SortingDirection;
+};
+
+export enum TableAction {
+  DEFAULT = 'default',
+  FIND_IN_TABLE = 'find_in_table',
+  FILTERS = 'filters',
+  SHOW_HIDE_COLUMNS = 'show_hide_columns',
+  SHOW_HIDE_ROWS = 'show_hide_rows',
+  DOWNLOAD_SHOWN_DATA = 'download_shown_data',
+  DOWNLOAD_ALL_DATA = 'download_all_data',
+  RESTORE_DEFAULTS = 'restore_defaults'
+}
+
+export type TableCellData = ReactNode;
+
+export type TableRowData = TableCellData[];
+
+export type TableData = TableRowData[];
+
+export type IndividualColumn = {
+  columnId: string;
+  isSortable?: boolean;
+  title: string;
+};
+
+export type Columns = IndividualColumn[];
+
+export type TableState = {
+  columns: Columns;
+  data: TableData;
+  rowsPerPage: number;
+  currentPageNumber: number;
+  searchText: string;
+  hiddenRowIds: number[];
+  hiddenColumnIds: number[];
+  selectedRowIds: number[];
+  selectedAction: TableAction;
+  sortedColumn: SortedColumn | null;
+  expandedRows: number[];
+  fixedHeader: boolean;
+};
+
+export const defaultTableState: TableState = {
+  columns: [],
+  data: [],
+  rowsPerPage: 100,
+  currentPageNumber: 1,
+  searchText: '',
+  hiddenRowIds: [],
+  hiddenColumnIds: [],
+  selectedRowIds: [],
+  selectedAction: TableAction.DEFAULT,
+  sortedColumn: null,
+  expandedRows: [],
+  fixedHeader: false
+};
+
+type SetRowsPerPageAction = {
+  type: 'set_rows_per_page';
+  payload: number;
+};
+type SetSearchTextAction = {
+  type: 'set_search_text';
+  payload: string;
+};
+type SetCurrentPageNumberAction = {
+  type: 'set_current_page_number';
+  payload: number;
+};
+type SetHiddenRowIdsAction = {
+  type: 'set_hidden_row_ids';
+  payload: number[];
+};
+type SetHiddenColumnIdsAction = {
+  type: 'set_hidden_column_ids';
+  payload: number[];
+};
+type SetSelectedRowIdsAction = {
+  type: 'set_selected_row_ids';
+  payload: number[];
+};
+type SetSelectedActionAction = {
+  type: 'set_selected_action';
+  payload: TableAction;
+};
+type SetSortedColumnAction = {
+  type: 'set_sorted_column';
+  payload: SortedColumn;
+};
+type SetExpandedRowsAction = {
+  type: 'set_expanded_rows';
+  payload: number[];
+};
+
+export type AllTableActions =
+  | SetRowsPerPageAction
+  | SetSearchTextAction
+  | SetCurrentPageNumberAction
+  | SetHiddenRowIdsAction
+  | SetHiddenColumnIdsAction
+  | SetSelectedRowIdsAction
+  | SetSelectedActionAction
+  | SetSortedColumnAction
+  | SetExpandedRowsAction;
+
+export const tableReducer = (state: TableState, action: AllTableActions) => {
+  switch (action.type) {
+    case 'set_rows_per_page':
+      return { ...state, rowsPerPage: action.payload };
+    case 'set_search_text':
+      return { ...state, searchText: action.payload };
+    case 'set_current_page_number':
```

Comment:
The problem here, I think, is that this will create two sources of truth that will need to be reconciled with each other.

One source of truth would be the parent's state, because the parent component (which is providing the data to the table) will need to be aware of the page, and of number of items per page, since it may use this to query a paginated backend api for data.

The reducer that you propose will be another piece of state kept within the table, and thus another source of truth.

This, I think, is dangerous. Which is why I think that it should not be the table's responsibility to store the state that its parent is also going to be interested in.

Reply:
>Which is why I think that it should not be the table's responsibility to store the state that its parent is also going to be interested in.

What if we don't want to retain the state of the table when we switch between apps? We would still need some sort of state within the table to handle actions like filtering, selecting and hiding rows. Should we force the parent component to be responsible to maintain the state even when we don't want to retain it?

I agree that it sounds dangerous to have two sources of truth but we only update the one within the table while we interact with it and update the parent state once before the unload or when needed (like informing a page change).